File: Source/Aladdin.ISO7816.Java/src/aladdin/iso7816/File.java
package aladdin.iso7816;
import aladdin.iso7816.ber.*; 
import java.io.*;

///////////////////////////////////////////////////////////////////////////////
// Файл
///////////////////////////////////////////////////////////////////////////////
public abstract class File 
{
    // родительский каталог и шаблон описания файла
    private final DedicatedFile parent; private final FileControlInformation info;
    
    // идентификатор, путь файла и способ кодирования данных
    private final Short id; private final short[] path; private final DataCoding dataCoding;
    
    // конструктор
    protected File(Short id, DataCoding dataCoding, FileControlInformation info) throws IOException
    { 
        // сохранить переданные параметры
        this.parent = null; this.info = info;

        // указать путь файла
        this.id = id; path = (id != null) ? new short[] { id } : null; 
        
        // сохранить способ кодирования данных
        this.dataCoding = this.info.getDataCoding(dataCoding); 
    }
    // конструктор
    protected File(DedicatedFile parent, Short id, FileControlInformation info) throws IOException 
    { 
        // сохранить переданные параметры
        this.parent = parent; this.info = info;

        // проверить наличие пути
        this.id = id; if (id == null || parent.path() == null) path = null; 
        
        // получить путь родительского каталога
        else { short[] parentPath = parent.path(); 
            
            // указать идентификатор файла
            path = new short[parentPath.length + 1]; path[parentPath.length] = id;
            
            // скопировать родительский путь
            System.arraycopy(parentPath, 0, path, 0, path.length - 1);
        }
        // сохранить способ кодирования данных
        this.dataCoding = this.info.getDataCoding(parent.dataCoding()); 
    }
    // каталог файла
    public final DedicatedFile parent() { return parent; } 
    
    // идентификатор файла
    public final Short id() { return id; } 
    // путь файла
    public final short[] path() { return path; }
    
    // способ кодирования данных
    public final DataCoding dataCoding() { return dataCoding; }
    // информация файла
    public final FileControlInformation info() { return info; }
    
    // категория файла
    public int fileCategory() { return FileCategory.UNKNOWN; } 
    // структура файла
    public FileStructure fileStructure() { return FileStructure.UNKNOWN; }
    
    // выделить родительский каталог
    public abstract DedicatedFile selectParent(LogicalChannel channel) throws IOException; 
}

File: Source/Aladdin.ISO7816/File.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace Aladdin.ISO7816
{
    ///////////////////////////////////////////////////////////////////////////////
    // Файл
    ///////////////////////////////////////////////////////////////////////////////
    public abstract class File 
    {
        // родительский каталог и шаблон описания файла
        private DedicatedFile parent; private BER.FileControlInformation info;
    
        // идентификатор, путь файла и способ кодирования данных
        private ushort? id; private ushort[] path; private DataCoding dataCoding;
    
        // конструктор
        protected File(ushort? id, DataCoding dataCoding, BER.FileControlInformation info) 
        { 
            // сохранить переданные параметры
            this.parent = null; this.info = info;

            // указать путь файла
            this.id = id; path = (id.HasValue) ? new ushort[] { id.Value } : null; 

            // сохранить способ кодирования данных
            this.dataCoding = this.info.GetDataCoding(dataCoding); 
        }
        // конструктор
        protected File(DedicatedFile parent, ushort? id, BER.FileControlInformation info) 
        { 
            // сохранить переданные параметры
            this.parent = parent; this.info = info;

            // проверить наличие пути
            this.id = id; if (!id.HasValue || parent.Path == null) path = null; 
        
            // получить путь родительского каталога
            else { ushort[] parentPath = parent.Path; 
            
                // указать идентификатор файла
                path = new ushort[parentPath.Length + 1]; path[parentPath.Length] = id.Value;
            
                // скопировать родительский путь
                Array.Copy(parentPath, 0, path, 0, path.Length - 1);
            }
            // сохранить способ кодирования данных
            this.dataCoding = this.info.GetDataCoding(parent.DataCoding); 
        }
        // каталог файла
        public DedicatedFile Parent { get { return parent; }}
    
        // идентификатор файла
        public ushort? ID { get { return id; }}
        // путь файла
        public ushort[] Path { get { return path; }}
    
        // информация файла
        public BER.FileControlInformation Info { get { return info; }}
        // способ кодирования данных
        public DataCoding DataCoding { get { return dataCoding; }}
    
        // категория файла
        public virtual int FileCategory { get { return ISO7816.FileCategory.Unknown; }}
        // структура файла
        public virtual FileStructure FileStructure { get { return ISO7816.FileStructure.Unknown; }}

        // выделить родительский каталог
        public abstract DedicatedFile SelectParent(LogicalChannel channel); 
    }
}
